src/metabase/util.clj

```diff
@@ -94,6 +94,21 @@
              (re-matches #"[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?"
                          (str/lower-case s)))))
 
+(defn state?
+  "Is `s` a state string?"
+  ^Boolean [^String s]
+  (contains? #{"alabama" "alaska" "arizona" "arkansas" "california" "colorado" "connecticut" "delaware"
+               "florida" "georgia" "hawaii" "idaho" "illinois" "indiana" "iowa" "kansas" "kentucky" "louisiana"
+               "maine" "maryland" "massachusetts" "michigan" "minnesota" "mississippi" "missouri" "montana"
+               "nebraska" "nevada" "new hampshire" "new jersey" "new mexico" "new york" "north carolina"
+               "north dakota" "ohio" "oklahoma" "oregon" "pennsylvania" "rhode island" "south carolina"
+               "south dakota" "tennessee" "texas" "utah" "vermont" "virginia" "washington" "west virginia"
+               "wisconsin" "wyoming"
+               "al" "mt" "ak" "ne" "az" "nv" "ar" "nh" "ca" "nj" "co" "nm" "ct" "ny" "de" "nc" "fl" "nd"
+               "ga" "oh" "hi" "ok" "id" "or" "il" "pa" "in" "ri" "ia" "sc" "ks" "sd" "ky" "tn" "la" "tx"
+               "me" "ut" "md" "vt" "ma" "va" "mi" "wa" "mn" "wv" "ms" "wi" "mo" "wy"}
+             (when s (str/lower-case s))))
```

Comment:
I'd lift the `when` to include `contains?` and/or use `some->>`.

Also not sure which is the better tradeoff: keeping the set in memory permanently, or paying the cost to (potentially) recreate it on every invocation. I'm leaning towards the former, but this should be tested.

Reply:
the clojure compiler will emit this into the constant pool i believe. its only made into a set on reading

Reply:
```clojure
(decompile (defn state?
  "Is `s` a state string?"
  ^Boolean [^String s]
  (contains? #{"alabama" "alaska" "arizona" "arkansas" "california" "colorado" "connecticut" "delaware"
               "florida" "georgia" "hawaii" "idaho" "illinois" "indiana" "iowa" "kansas" "kentucky" "louisiana"
               "maine" "maryland" "massachusetts" "michigan" "minnesota" "mississippi" "missouri" "montana"
               "nebraska" "nevada" "new hampshire" "new jersey" "new mexico" "new york" "north carolina"
               "north dakota" "ohio" "oklahoma" "oregon" "pennsylvania" "rhode island" "south carolina"
               "south dakota" "tennessee" "texas" "utah" "vermont" "virginia" "washington" "west virginia"
               "wisconsin" "wyoming"
               "al" "mt" "ak" "ne" "az" "nv" "ar" "nh" "ca" "nj" "co" "nm" "ct" "ny" "de" "nc" "fl" "nd"
               "ga" "oh" "hi" "ok" "id" "or" "il" "pa" "in" "ri" "ia" "sc" "ks" "sd" "ky" "tn" "la" "tx"
               "me" "ut" "md" "vt" "ma" "va" "mi" "wa" "mn" "wv" "ms" "wi" "mo" "wy"}
             (when s (str/lower-case s)))))

// Decompiling class: user$state_QMARK_
import clojure.lang.*;

public final class user$state_QMARK_ extends AFunction
{
    public static final Var const__0;
    public static final AFn const__1;
    public static final Var const__2;

    public static Object invokeStatic(final Object s) {
        final IFn fn = (IFn)user$state_QMARK_.const__0.getRawRoot();
        final AFn const__1 = user$state_QMARK_.const__1;
        if (s != null) {
            if (s != Boolean.FALSE) {
                final Object invoke = ((IFn)user$state_QMARK_.const__2.getRawRoot()).invoke(s);
                return fn.invoke(const__1, invoke);
            }
        }
        final Object invoke = null;
        return fn.invoke(const__1, invoke);
    }

    @Override
    public Object invoke(final Object s) {
        return invokeStatic(s);
    }

    static {
        const__0 = RT.var("clojure.core", "contains?");
        const__1 = PersistentHashSet.create("rhode island", "delaware", "ct", "ohio", "new york", "colorado", "nh", "wi", "la", "california", "md", "wisconsin", "massachusetts", "ne", "alaska", "al", "pennsylvania", "oklahoma", "texas", "mo", "vt", "north carolina", "idaho", "nevada", "new jersey", "indiana", "west virginia", "florida", "new mexico", "nc", "ky", "maine", "sc", "kentucky", "id", "arkansas", "utah", "co", "connecticut", "vermont", "ca", "mi", "ut", "wv", "kansas", "ma", "ia", "wa", "ms", "south dakota", "fl", "il", "oh", "south carolina", "tn", "or", "de", "ri", "hi", "wy", "iowa", "maryland", "ga", "hawaii", "montana", "wyoming", "mt", "tx", "missouri", "oregon", "georgia", "nm", "mn", "north dakota", "virginia", "illinois", "ar", "ny", "sd", "ak", "ok", "alabama", "nv", "louisiana", "nebraska", "tennessee", "new hampshire", "washington", "pa", "va", "minnesota", "ks", "me", "nj", "arizona", "michigan", "in", "az", "mississippi", "nd");
        const__2 = RT.var("clojure.string", "lower-case");
    }
}

nil
```

yeah. and checking for nil versus checking the set for nil is probably not a huge difference.

Reply:
Good catch. TIL -- I thought it only does that for small objects.